src/abstractdataframe/io.jl

```diff
@@ -217,3 +217,89 @@ end
+function Data.schema(df::DataFrame, ::Type{Data.Column})
+    return Data.Schema(map(string, names(df)),
+                       DataType[typeof(A) for A in df.columns], size(df, 1))
+end
+
+# DataFrame as a Data.Source
+function Data.isdone(source::DataFrame, row, col)
+    rows, cols = size(source)
+    return row > rows || col > cols
+end
+
+Data.streamtype(::Type{DataFrame}, ::Type{Data.Column}) = true
+Data.streamtype(::Type{DataFrame}, ::Type{Data.Field}) = true
+
+Data.streamfrom{T <: AbstractVector}(source::DataFrame, ::Type{Data.Column}, ::Type{T}, col) =
+    (@inbounds A = source.columns[col]::T; return A)
+Data.streamfrom{T}(source::DataFrame, ::Type{Data.Column}, ::Type{T}, col) =
+    (@inbounds A = source.columns[col]; return A)
+Data.streamfrom{T}(source::DataFrame, ::Type{Data.Field}, ::Type{T}, row, col) =
+    (@inbounds A = Data.streamfrom(source, Data.Column, T, col); return A[row]::T)
+
+# DataFrame as a Data.Sink
+allocate{T}(::Type{T}, rows) = Vector{T}(rows)
+allocate{T}(::Type{Vector{T}}, rows) = Vector{T}(rows)
+allocate{T}(::Type{Nullable{T}}, rows) = DataArray(Vector{T}(rows))
+allocate{T}(::Type{DataVector{T}}, rows) = DataArray(Vector{T}(rows))
+allocate{S,R}(::Type{PooledDataVector{S,R}}, rows) = PooledDataArray{S,1,R}(rows)
+
+function DataFrame{T <: Data.StreamType}(sch::Data.Schema,
+                                         ::Type{T}=Data.Field,
+                                         append::Bool=false,
+                                         ref::Vector{UInt8}=UInt8[], args...)
+    rows, cols = size(sch)
+    rows = max(0, T <: Data.Column ? 0 : rows) # don't pre-allocate for Column streaming
+    columns = Vector{Any}(cols)
+    types = Data.types(sch)
+    for i = 1:cols
+        columns[i] = allocate(types[i], rows)
+    end
+    return DataFrame(columns, map(Symbol, Data.header(sch)))
+end
+
+# given an existing DataFrame (`sink`), make any necessary changes for streaming source
+# with Data.Schema `sch` to it, given we know if we'll be `appending` or not
+function DataFrame(sink, sch::Data.Schema, ::Type{Data.Field}, append::Bool, ref::Vector{UInt8})
+    rows, cols = size(sch)
+    newsize = max(0, rows) + (append ? size(sink, 1) : 0)
+    newsize != size(sink, 1) && foreach(x->resize!(x, newsize), sink.columns)
+    sch.rows = newsize
+    return sink
+end
+function DataFrame(sink, sch::Data.Schema, ::Type{Data.Column}, append::Bool, ref::Vector{UInt8})
+    rows, cols = size(sch)
+    append ? (sch.rows += size(sink, 1)) : foreach(empty!, sink.columns)
+    return sink
+end
+
+Data.streamtypes(::Type{DataFrame}) = [Data.Column, Data.Field]
+
+Data.streamto!{T}(sink::DataFrame, ::Type{Data.Field}, val::T, row, col, sch::Data.Schema{false}) =
+    push!(sink.columns[col]::Vector{T}, val)
+Data.streamto!{T}(sink::DataFrame, ::Type{Data.Field}, val::Nullable{T}, row, col, sch::Data.Schema{false}) =
+    push!(sink.columns[col]::DataVector{T}, isnull(val) ? NA : get(val))
```

Comment:
`get(val, NA)` should work and be more efficient (since internally it uses `ifelse` to avoid a branch when possible). Same below.